Create SignalR & BackgroundService article (#10756)
aspnet/Docs: aspnetcore/signalr/background-services.md

---
title: SignalR with Background Services
author: bradygaster
description: Learn how to send messages to SignalR clients from .NET Core BackgroundService classes.
monikerRange: '>= aspnetcore-2.2'
ms.author: bradyg
ms.custom: mvc
ms.date: 01/31/2018
uid: signalr/background-services
---
# Hosting SignalR in Background Services

By [Brady Gaster](https://twitter.com/bradygaster)

> This article provides guidance for hosting SignalR Hubs using background worker process hosted with ASP.NET Core, and how to send messages to connected clients from within a .NET Core [BackgroundService](/dotnet/api/microsoft.extensions.hosting.backgroundservice).

Reviewer: ```suggestion
This article provides guidance for:

* Hosting SignalR Hubs using a background worker process hosted with ASP.NET Core.
* Sending messages to connected clients from within a .NET Core [BackgroundService](xref:Microsoft.Extensions.Hosting.BackgroundService).
```

[View or download sample code](https://github.com/aspnet/Docs/tree/master/aspnetcore/signalr/background-services/sample/) [(how to download)](xref:index#how-to-download-a-sample)

> ## Wiring up SignalR during Startup

Reviewer: ```suggestion
## Wire up SignalR during startup
```

> Hosting ASP.NET Core SignalR Hubs in the context of a background worker process is identical to that of hosting a Hub in an ASP.NET Core web app. In the `ConfigureServices` method of a `Startup` class, calling `services.AddSignalR` adds the required services to the ASP.NET Core DI layer to support SignalR. Then, in `Configure`, the `UseSignalR` method is called to wire up the Hub endpoint(s) in the ASP.NET Core request pipeline.

Reviewer: ```suggestion
Hosting ASP.NET Core SignalR Hubs in the context of a background worker process is identical to that of hosting a Hub in an ASP.NET Core web app. In the `Startup.ConfigureServices` method, calling `services.AddSignalR` adds the required services to the ASP.NET Core Dependency Injection (DI) layer to support SignalR. In `Startup.Configure`, the `UseSignalR` method is called to wire up the Hub endpoint(s) in the ASP.NET Core request pipeline.
```

[!code-csharp[Startup](background-service/sample/Server/Startup.cs?name=Startup)]

> In the example code, the `ClockHub` class implements the `Hub<T>` class to create a strongly-typed Hub. The `ClockHub` has been configured in the `Startup` class to respond to requests at the endpoint `/hubs/clock`.

Reviewer: ```suggestion
In the preceding example, the `ClockHub` class implements the `Hub<T>` class to create a strongly typed Hub. The `ClockHub` has been configured in the `Startup` class to respond to requests at the endpoint `/hubs/clock`.
```

Author: i tried to avoid terms like "preceding" and "below" per feedback i've consistently received from @tdykstra and switched to using "example code" in most cases. Is there a convention for referring to the example code?

> > For more information on strongly-typed Hubs, see the article [Use hubs in SignalR for ASP.NET Core](/aspnet/core/signalr/hubs?view=aspnetcore-2.2#strongly-typed-hubs).

Reviewer: Make this sentence part of the previous paragraph.

Prefer xref-style links, since those are validated during the build process. The value following `xref:` is the `uid` value in the targeted doc's metadata.

```suggestion
> For more information on strongly typed Hubs, see [Use hubs in SignalR for ASP.NET Core](xref:signalr/hubs#strongly-typed-hubs).
```

Author: ah, the xref. i will refer to this in the future as a reference on how to use... uh... *references*. 👍

[!code-csharp[Startup](background-service/sample/Server/ClockHub.cs?name=ClockHub)]

> The interface used by the strongly-typed `ClockHub` is the `IClock` interface.

Reviewer: ```suggestion
The interface used by the strongly typed `ClockHub` is the `IClock` interface.
```

[!code-csharp[Startup](background-service/sample/HubServiceInterfaces/IClock.cs?name=IClock)]

> ## Calling a SignalR Hub from a background service

Reviewer: ```suggestion
## Call a SignalR Hub from a background service
```

During startup, the `Worker` class, a `BackgroundService`, is wired up using `AddHostedService`.

```csharp
services.AddHostedService<Worker>();
```

> Since SignalR is also wired up during the `Startup` phase in which each Hub is attached to an individual endpoint in ASP.NET Core's HTTP request pipeline, each Hub is represented by an `IHubContext<T>` on the server. Using ASP.NET Core's dependency injection features, other classes instantiated by the hosting layer - like `BackgroundService` classes or MVC Controller classes or even Razor page models - can get references to server-side Hubs by accepting instances of  `IHubContext<ClockHub, IClock>` during construction.

Reviewer: ```suggestion
Since SignalR is also wired up during the `Startup` phase, in which each Hub is attached to an individual endpoint in ASP.NET Core's HTTP request pipeline, each Hub is represented by an `IHubContext<T>` on the server. Using ASP.NET Core's DI features, other classes instantiated by the hosting layer, like `BackgroundService` classes, MVC Controller classes, or Razor page models, can get references to server-side Hubs by accepting instances of `IHubContext<ClockHub, IClock>` during construction.
```

[!code-csharp[Startup](background-service/sample/Server/Worker.cs?name=Worker)]

> As the `ExecuteAsync` method is called iteratively in the background service, the server's current date & time are sent to the connected clients using the `ClockHub`.

Reviewer: ```suggestion
As the `ExecuteAsync` method is called iteratively in the background service, the server's current date and time are sent to the connected clients using the `ClockHub`.
```

> ## Reacting to SignalR events with Background Services

Reviewer: ```suggestion
## React to SignalR events with background services
```

Author: If i had referred to the class name, would you have asked me to leave it as BackgroundServices? Or, had i done that, would the recommendation had been to split it up and make it lower-case? Just curious.

> Using the [.NET Client for SignalR](/aspnet/core/signalr/dotnet-client), a `BackgroundService` or `IHostedService` implementor can also be used to connect to SignalR Hubs and respond to events, much like HTML or .NET client applications do using the JavaScript and .NET Clients for SignalR.

Reviewer: ```suggestion
Using the <xref:signalr/dotnet-client>, a `BackgroundService` or `IHostedService` implementor can also be used to connect to SignalR Hubs and respond to events, much like HTML or .NET client apps do using the JavaScript and .NET clients for SignalR.
```

The `ClockHubClient` class implements both the `IClock` interface and the `IHostedService` interface. This way it can be wired up during `Startup` to run continuously and respond to Hub events from the server. 

```csharp
public partial class ClockHubClient : IClock, IHostedService
{
}
```

During initialization, the `ClockHubClient` creates an instance of a `HubConnection` and wires up the `IClock.ShowTime` method as the handler for the Hub's `ShowTime` event.

[!code-csharp[The ClockHubClient constructor](background-service/sample/Clients.ConsoleTwo/ClockHubClient.cs?name=ClockHubClientCtor)]

In the `IHostedService.StartAsync` implementation, the `HubConnection` is started asynchronously.

[!code-csharp[StartAsync method](background-service/sample/Clients.ConsoleTwo/ClockHubClient.cs?name=StartAsync)]

During the `IHostedService.StopAsync` method, the `HubConnection` is disposed of asynchronously.

[!code-csharp[StopAsync method](background-service/sample/Clients.ConsoleTwo/ClockHubClient.cs?name=StopAsync)]

> ## Related resources

Reviewer: ```suggestion
## Additional resources
```

* [Get started](xref:tutorials/signalr)
* [Hubs](xref:signalr/hubs)
* [Publish to Azure](xref:signalr/publish-to-azure-web-app)

> * [Strongly-typed Hubs](/aspnet/core/signalr/hubs?view=aspnetcore-2.2#strongly-typed-hubs)

Reviewer: ```suggestion
* [Strongly typed Hubs](xref:signalr/hubs#strongly-typed-hubs)
```

* Hosting SignalR Hubs using a background worker process hosted with ASP.NET Core.
* Sending messages to connected clients from within a .NET Core [BackgroundService](xref:Microsoft.Extensions.Hosting.BackgroundService).

[View or download sample code](https://github.com/aspnet/Docs/tree/master/aspnetcore/signalr/background-services/sample/) [(how to download)](xref:index#how-to-download-a-sample)

## Wire up SignalR during startup

Hosting ASP.NET Core SignalR Hubs in the context of a background worker process is identical to that of hosting a Hub in an ASP.NET Core web app. In the `Startup.ConfigureServices` method, calling `services.AddSignalR` adds the required services to the ASP.NET Core Dependency Injection (DI) layer to support SignalR. In `Startup.Configure`, the `UseSignalR` method is called to wire up the Hub endpoint(s) in the ASP.NET Core request pipeline.

[!code-csharp[Startup](background-service/sample/Server/Startup.cs?name=Startup)]

In the preceding example, the `ClockHub` class implements the `Hub<T>` class to create a strongly typed Hub. The `ClockHub` has been configured in the `Startup` class to respond to requests at the endpoint `/hubs/clock`.

> > For more information on strongly typed Hubs, see [Use hubs in SignalR for ASP.NET Core](xref:signalr/hubs#strongly-typed-hubs).

Reviewer: ```suggestion
For more information on strongly typed Hubs, see [Use hubs in SignalR for ASP.NET Core](xref:signalr/hubs#strongly-typed-hubs).
```

---
title: Host ASP.NET Core SignalR in background services
author: bradygaster
description: Learn how to send messages to SignalR clients from .NET Core BackgroundService classes.
monikerRange: '>= aspnetcore-2.2'
ms.author: bradyg
ms.custom: mvc

> ms.date: 01/31/2018

Reviewer 2: 2019

ms.date: 01/31/2018
uid: signalr/background-services
---
# Host ASP.NET Core SignalR in background services

By [Brady Gaster](https://twitter.com/bradygaster)

This article provides guidance for:

* Hosting SignalR Hubs using a background worker process hosted with ASP.NET Core.
* Sending messages to connected clients from within a .NET Core [BackgroundService](xref:Microsoft.Extensions.Hosting.BackgroundService).

[View or download sample code](https://github.com/aspnet/Docs/tree/master/aspnetcore/signalr/background-services/sample/) [(how to download)](xref:index#how-to-download-a-sample)

## Wire up SignalR during startup

> Hosting ASP.NET Core SignalR Hubs in the context of a background worker process is identical to that of hosting a Hub in an ASP.NET Core web app. In the `Startup.ConfigureServices` method, calling `services.AddSignalR` adds the required services to the ASP.NET Core Dependency Injection (DI) layer to support SignalR. In `Startup.Configure`, the `UseSignalR` method is called to wire up the Hub endpoint(s) in the ASP.NET Core request pipeline.

Reviewer 3: ~~that of~~

The interface used by the strongly typed `ClockHub` is the `IClock` interface.

[!code-csharp[Startup](background-service/sample/HubServiceInterfaces/IClock.cs?name=IClock)]

## Call a SignalR Hub from a background service

During startup, the `Worker` class, a `BackgroundService`, is wired up using `AddHostedService`.

```csharp
services.AddHostedService<Worker>();
```

Since SignalR is also wired up during the `Startup` phase, in which each Hub is attached to an individual endpoint in ASP.NET Core's HTTP request pipeline, each Hub is represented by an `IHubContext<T>` on the server. Using ASP.NET Core's DI features, other classes instantiated by the hosting layer, like `BackgroundService` classes, MVC Controller classes, or Razor page models, can get references to server-side Hubs by accepting instances of `IHubContext<ClockHub, IClock>` during construction.

[!code-csharp[Startup](background-service/sample/Server/Worker.cs?name=Worker)]

As the `ExecuteAsync` method is called iteratively in the background service, the server's current date and time are sent to the connected clients using the `ClockHub`.

## React to SignalR events with background services

> Using the <xref:signalr/dotnet-client>, a `BackgroundService` or `IHostedService` implementor can also be used to connect to SignalR Hubs and respond to events, much like HTML or .NET client apps do using the JavaScript and .NET clients for SignalR.

Reviewer 3: This sentence is hard to follow:
* Should "implementor" be "implementation"?
* Using the .NET client ... much like ... .NET client apps do ... using .NET clients [sounds circular]
* Could "can also be used to connect" be changed to "can connect"?
* "like HTML or .NET client apps do using the JavaScript and .NET clients for SignalR"
  * confusing combination of "or" and "and"
  * does this mean to say "like an HTML page can do by using the JavaScript client, and like a .NET desktop app can do by using the .NET client"?

## React to SignalR events with background services

> Like an HTML using the JavaScript client for SignalR or a .NET desktop app can do using the using the <xref:signalr/dotnet-client>, a `BackgroundService` or `IHostedService` implementation can also be used to connect to SignalR Hubs and respond to events.

Reviewer: I suggest changing this first sentence to begin with "Like a Single Page App using..." instead of "Like an HTML using...".
